### ISSUE TITLE
Render EditorComponent on main site

### ISSUE THREAD

**Issue by jozsi** [netlify/netlify-cms#3720]
Hello,

I've registered a couple of editor components for the markdown widget that work fine in the preview pane of the Netlify CMS admin panel. However, I was hoping to see the same result on the actual site as well, supposing that either `gatsby-plugin-netlify` or `gatsby-plugin-netlify-cms` would handle that for me, but I realized it's out of their scope.

So the only way I - as a beginner to Gatsby - could work this around, is to use `rehype-react` and a shared component between it and `registerEditorComponent's toPreview`.

I am wondering if there is a more elegant solution to do this, until the MDX support gets ready?

Thank you!

**Comment by HeyITGuyFixIt** [comment]
What code do you have for your editor components? Be sure to set the toBlock and the pattern parameters appropriately. The content that goes into toBlock needs to be what goes into the actual file being edited. And the pattern needs to be modified to match the output of toBlock. Simply settting toPreview won't work.

I found this out the hard way, and it took a bit of experimenting to find out. The documentation is really not clear on this.

**Comment by jozsi** [comment]
Yep, in the admin panel everything works flawlessly, back and forward. I thought the pattern would be picked up when generating the static pages. Cause what I save in the markdown is a shortcode wannabe, not markdown, not HTML.

**Comment by HeyITGuyFixIt** [comment]
Can you share an example of one of your components?

**Comment by jozsi** [comment]
Sure. I want a button widget. My first attempt was like this (simplified):
```js
import React from 'react';
import CMS from 'netlify-cms-app';

function ButtonComponent({ label, href }) {
  return (
    <a href={href} class="button is-primary">
      {label}
    </a>
  );
}

CMS.registerEditorComponent({
  id: 'button',
  label: 'Button',
  fields: [
    { name: 'label', label: 'Label', widget: 'string' },
    { name: 'href', label: 'Link', widget: 'string' },
  ],
  pattern: /^\((.+)\)\[(.+)\]$/,
  fromBlock: function(match) {
    return {
      label: match[1],
      href: match[2],
    };
  },
  toBlock: function(obj) {
    return `(${obj.label})` + (obj.href ? `[${obj.href}]` : '');
  },
  toPreview: ButtonComponent,
});
```
It saved `(Button Label)[Button Link]` in the .md file rendered a HTML button in the preview, but on the site was just that static text.

So my next attempt was to update the `pattern` and `toBlock` of the CMS widget to work with the following:
```
<my-button-component href="https://netlify.com">Button Label</my-button-component>
```
and use [react-rehype as described here](https://using-remark.gatsbyjs.org/custom-components/) to render the Button component (basically replace `html` with `htmlAst` in the GraphQL query and use:
```js
const renderAst = new rehypeReact({
  createElement: React.createElement,
  components: { 'my-button-component': ButtonComponent },
}).Compiler;

export const HTMLContent = ({ content, className }) => (
  <div className={className}>{renderAst(content)}</div>
);
```

I can provide the full code to whoever is interested, but it got really complex to use as an example.

**Comment by HeyITGuyFixIt** [comment]
What does Gatsby use for converting MD files? It seems like something is wrong with that, as you said Netlify CMS is correctly adding the code to the file. Check with the proper syntax Gatsby expects for links to see if it matches what you are using. It looks to me like Gatsby doesn't recognize it.

**Comment by HeyITGuyFixIt** [comment]
Answering my own question (lol). See [here](https://www.gatsbyjs.org/docs/mdx/markdown-syntax/#link) for the proper link syntax. It looks like you got it backwards. 

It is supposed to be `[text](url)`, while you said you have (text)[url]`. I have made the same mistake multiple times before. I almost made it again with the link above.

**Comment by jozsi** [comment]
The inverse pattern was intentional, trying to keep somewhat a consitency between a link and a button. However, I've read about some issues and it is better to use backticks or curly brackets for custom stuff.

Gatsby uses Remark for converting MD files. Haven't had time to jump in the code see how Netlify CMS tackles it - if anyone is open to give a brief info about it, I'm eager to hear!

**Comment by smashercosmo** [comment]
Yes, every time you write custom editor component, you should write corresponding remark plugin. Example:

video-player-editor-component.js
```js
export const VideoPlayer = {
  id: 'video-player',
  label: 'Video Player',
  fields: [{ name: 'src', label: 'Video URL', widget: 'file' }],
  pattern: /^video-player:(\S+)$/,
  fromBlock(match: RegExpMatchArray) {
    return {
      src: match[1],
    }
  },
  toBlock(obj: { src: string }) {
    return `video-player:${obj.src}`
  },
  toPreview(obj: { src: string }) {
    if (!obj.src) return ''

    return `
      <div style="padding-bottom: 56.25%; position: relative; height: 0; overflow: hidden;">
        <video
          style="position: absolute; top: 0; left: 0; width: 100%; height: 100%;"
          autoPlay
          muted
          loop
          playsInline>
          <source src="${obj.src}" type="video/mp4" />
        </video>
      </div>
    `
  },
}
```

video-player-remark-plugin.js
```js
const path = require('path')
const visit = require('unist-util-visit')

module.exports = ({ markdownAST }) => {
  visit(markdownAST, 'paragraph', (node) => {
    const { children } = node
    const [mayBeText] = children

    if (
      !mayBeText ||
      mayBeText.type !== 'text' ||
      mayBeText.value.indexOf('video-player:') !== 0
    ) {
      return
    }

    const [, link] = mayBeText.value.split(':')
    if (path.extname(link) === '.mp4') {
      delete node.children
      node.type = 'html'
      node.value = `
        <div style="padding-bottom: 56.25%; position: relative; height: 0; overflow: hidden;">
          <video
            style="position: absolute; top: 0; left: 0; width: 100%; height: 100%;"
            autoPlay
            muted
            loop
            playsInline>
            <source src="${link}" type="video/mp4" />
          </video>
        </div>
      `
    }
  })
  return markdownAST
}
```
